src/middleware/messageStore.js

```diff
@@ -26,132 +26,280 @@ function copyMapWithEscapedReservedCharacters (map) {
+        port: (ctx.host != null ? ctx.host.split(':')[1] : undefined),
+        path: ctx.path,
+        headers,
+        querystring: ctx.querystring,
+        method: ctx.method,
+        timestamp: ctx.requestTimestamp
+      }
+    })
+
+    if (ctx.parentID && ctx.taskID) {
+      tx.parentID = ctx.parentID
+      tx.taskID = ctx.taskID
+    }
+
+    if (ctx.currentAttempt) {
+      tx.autoRetryAttempt = ctx.currentAttempt
+    }
+
+    // check if channel request body is false and remove - or if request body is empty
+    if ((ctx.authorisedChannel && ctx.authorisedChannel.requestBody === false) || (tx.request.body === '')) {
+      // reset request body
+      ctx.body = ''
+      // check if method is POST|PUT|PATCH - rerun not possible without request body
+      if (['POST', 'PUT', 'PATCH'].includes(ctx.method)) {
+        tx.canRerun = false
+      }
+    }
+
+    tx.save((err, tx) => {
+      if (err) {
+        logger.error(`Could not save transaction metadata (initiateRequest): ${err}`)
+        reject(err)
+      } else {
+        ctx.transactionId = tx._id
+        ctx.header['X-OpenHIM-TransactionID'] = tx._id.toString()
+        logger.info(`Done initiateRequest for transaction: ${tx._id}`)
+        resolve(tx)
+      }
+    })
   })
+}
 
-  if (ctx.parentID && ctx.taskID) {
-    tx.parentID = ctx.parentID
-    tx.taskID = ctx.taskID
-  }
+/*
+ *  Find and update an existing transaction once a Request has completed streaming
+ *    into the HIM (Not async; Mongo should handle locking issues, etc)
+ */
+export function completeRequest (ctx, done) {
 
-  if (ctx.currentAttempt) {
-    tx.autoRetryAttempt = ctx.currentAttempt
+  if (ctx && !ctx.requestTimestampEnd) {
+    ctx.requestTimestampEnd = new Date()
   }
 
-  // check if channel request body is false and remove - or if request body is empty
-  if ((ctx.authorisedChannel.requestBody === false) || (tx.request.body === '')) {
-    // reset request body
-    tx.request.body = ''
-    // check if method is POST|PUT|PATCH - rerun not possible without request body
-    if ((ctx.method === 'POST') || (ctx.method === 'PUT') || (ctx.method === 'PATCH')) {
-      tx.canRerun = false
+  const transactionId = getTransactionId(ctx)
+
+  return transactions.TransactionModel.findById(transactionId, (err, tx) => {
+    if (err) { return done(err) }
+
+    if (!tx) {
+      const errorMessage = `Could not find transaction with id ${transactionId}`
+      logger.error(errorMessage)
+      return done(new Error(errorMessage))
     }
-  }
 
-  if (utils.enforceMaxBodiesSize(ctx, tx.request)) { tx.canRerun = false }
+    /*
+     *  For short transactions, the 'end' timestamp is before the 'start' timestamp.
+     *  (Persisting the transaction initially takes longer than fully processing the transaction)
+     *  In these cases, swop the 'start' and 'end' values around; the transaction duration is
+     *  not exactly accurate, but at least it isn't negative.
+     */
```

Comment:
This probably means that we aren't capturing the start timestamp in the right place. Rather than this hack we should try find a reliable place to capture the start timestamp.